### DISCUSSION TITLE
Using Unity Package Manager with renovate

### DISCUSSION THREAD

**Discussion by DevHazzardt** [renovatebot/renovate#19300] [Request Help] [answered]
Hello! I am brand new to renovate and just had a quick question. I was doing some investigation but I can't see any way of using the unity package manager with renovate and I couldn't find any previous discussions surrounding this. I was wondering if there is any way for me to do this that isn't particularly obvious so I could use a manager in tandem with something like the manifest.json file unity uses.

**Comment by rarkins** [comment, marked as answer]
The key questions are:
- How complex is the file format which describes the dependencies?
- What type of dependencies does it reference? e.g. npm packages, or github repos?

**Reply by DevHazzardt** [reply]
Ah I see since they are npm packages I could use this.

**Reply by rarkins** [reply]
In that case you'd need to use `regexManagers` to custom parse/extract the unity file(s) and then the existing npm versioning and datasource may be all you need

**Reply by DevHazzardt** [reply]
That makes total sense. Thanks for clarifying so fast!

**Comment by bdovaz** [comment]
@rarkins I managed to make a rule for official Unity packages:

```json
{
    "matchDatasources": [
        "npm"
    ],
    "matchPackageNames": [
        "/^com.unity/"
    ],
    "registryUrls": [ "https://packages.unity.com" ]
}
```

But the problem is that Unity has the concept of `Scoped registries` (= custom registries) that in the following link you can see an example below: https://docs.unity3d.com/Manual/upm-scoped.html

It is the equivalent of `Package Source Mapping` of NuGet: https://learn.microsoft.com/en-us/nuget/consume-packages/package-source-mapping#enable-by-manually-editing-nugetconfig

The problem is that if there is nothing in Renovate that reads `Scoped registries` from each `manifest.json` of each Unity project we would have to keep N rules by hand that are not synchronized with the registries of the manifest file since we can add or remove in one file and forget to do it in the other.

@rarkins I have no problem in creating PR but I am not clear what I should create a PR from? What fits to solve this problem? datasource, manager, ...

More information:

**Project manifest:** https://docs.unity3d.com/Manual/upm-manifestPrj.html

**Package manifest:** https://docs.unity3d.com/Manual/upm-manifestPkg.html

**Reply by bdovaz** [reply]
Datasource created: https://github.com/renovatebot/renovate/pull/35298
